stats/opentelemetry/internal/grpc_trace_bin_propagator.go

```diff
@@ -0,0 +1,140 @@
+// binary format using the 'grpc-trace-bin' header (fast path). Otherwise,
+// the trace data is base64 encoded and injected using the same header in
+// text format (slow path).
+func (p GRPCTraceBinPropagator) Inject(ctx context.Context, carrier otelpropagation.TextMapCarrier) {
+	span := oteltrace.SpanFromContext(ctx)
+	if !span.SpanContext().IsValid() {
+		return
+	}
+
+	bd := Binary(span.SpanContext())
+	if bd == nil {
+		return
+	}
+
+	if cc, ok := carrier.(*itracing.CustomCarrier); ok {
+		cc.SetBinary(bd)
+		return
+	}
+	carrier.Set(itracing.GRPCTraceBinHeaderKey, base64.StdEncoding.EncodeToString(bd))
+}
+
+// Extract reads OpenTelemetry trace context information from the carrier into a
+// Context.
+//
+// If the carrier is a CustomCarrier, trace data is read directly in a binary
+// format from the 'grpc-trace-bin' header (fast path). Otherwise, the trace
+// data is base64 decoded from the same header in text format (slow path).
+func (p GRPCTraceBinPropagator) Extract(ctx context.Context, carrier otelpropagation.TextMapCarrier) context.Context {
+	var bd []byte
+
+	if cc, ok := carrier.(*itracing.CustomCarrier); ok {
+		bd = cc.GetBinary()
+	} else {
+		bd, _ = base64.StdEncoding.DecodeString(carrier.Get(itracing.GRPCTraceBinHeaderKey))
+	}
+	if bd == nil {
+		return ctx
+	}
+
+	sc, ok := FromBinary([]byte(bd))
+	if !ok {
+		return ctx
+	}
+
+	return oteltrace.ContextWithRemoteSpanContext(ctx, sc)
+}
+
+// Fields returns the keys whose values are set with Inject.
+//
+// GRPCTraceBinPropagator always returns a slice containing only
+// `grpc-trace-bin` key because it only sets the 'grpc-trace-bin' header for
+// propagating trace context.
+func (p GRPCTraceBinPropagator) Fields() []string {
+	return []string{itracing.GRPCTraceBinHeaderKey}
+}
+
+// Binary returns the binary format representation of a SpanContext.
+//
+// If sc is the zero value, returns nil.
+func Binary(sc oteltrace.SpanContext) []byte {
+	if sc.Equal(oteltrace.SpanContext{}) {
+		return nil
+	}
+	var b [29]byte
+	traceID := oteltrace.TraceID(sc.TraceID())
+	copy(b[2:18], traceID[:])
+	b[18] = 1
+	spanID := oteltrace.SpanID(sc.SpanID())
+	copy(b[19:27], spanID[:])
+	b[27] = 2
+	b[28] = uint8(oteltrace.TraceFlags(sc.TraceFlags()))
+	return b[:]
+}
+
+// FromBinary returns the SpanContext represented by b.
+//
+// If b has an unsupported version ID or contains no TraceID, FromBinary
+// returns with zero value SpanContext and false.
+func FromBinary(b []byte) (oteltrace.SpanContext, bool) {
+	if len(b) == 0 || b[0] != 0 {
+		return oteltrace.SpanContext{}, false
+	}
+	b = b[1:]
+	if len(b) < 17 || b[0] != 0 {
+		return oteltrace.SpanContext{}, false
+	}
+
+	sc := oteltrace.SpanContext{}
+	sc = sc.WithTraceID(oteltrace.TraceID(b[1:17]))
+	b = b[17:]
+	if len(b) >= 9 && b[0] == 1 {
+		sc = sc.WithSpanID(oteltrace.SpanID(b[1:9]))
+		b = b[9:]
+	}
+	if len(b) >= 2 && b[0] == 2 {
+		sc = sc.WithTraceFlags(oteltrace.TraceFlags(b[1]))
+	}
```

Comment:
This doesn't follow the C way and I think this is incorrect. In C, if the 18th index isn't 1 or the 27th index isn't 2 C errors. Even if there is a syntax error here, you just leave the array as zero values and return it with present zero values. That is incorrect, as only part of the returned SpanContext will be populated. It should be atomic, either it returns it or not, similar in philosophy as to why we gate concurrent reads and writes with a mutex so you don't get interspliced data as currently happens in your returned SpanContext.

Reply:
> if the 18th index isn't 1 or the 27th index isn't 2 C errors.

Below part take care of that.  it only parses, if 18th value is 1 and 27th value is 2. Its just the implementation reduces the array after parsing the part like `b = b[17:]` and then we check for `b[0]`
```
b = b[17:]
	if len(b) >= 9 && b[0] == 1 {
		sc = sc.WithSpanID(oteltrace.SpanID(b[1:9]))
		b = b[9:]
	}
	if len(b) >= 2 && b[0] == 2 {
		sc = sc.WithTraceFlags(oteltrace.TraceFlags(b[1]))
	}	
```

Reply:
> Even if there is a syntax error here, you just leave the array as zero values and return it with present zero values. That is incorrect, as only part of the returned SpanContext will be populated. It should be atomic, either it returns it or not, similar in philosophy as to why we gate concurrent reads and writes with a mutex so you don't get interspliced data as currently happens in your returned SpanContext.
 
 Again, since we are taking the code directly from an existing library I don't want to change much here as we know it has been a working code